_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+
+```julia
+using BenchmarkTools
+terrible_norm(x::AbstractVector{<:Real}) = only(reshape(x, 1, :) * x)
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+terrible_norm (generic function with 1 method)
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+terrible_norm(1:10)
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+385
+```
+
+</div>
+
+The thing that can go wrong here is that there are many kinds of element types that might only hold real numbers, but that do not have that as a type parameter.
+
+For example, the if the data ever contained `missing` values (common in data science),
+but you filtered them out somehow, the array will still be typed `Union{Missing, T}`
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+data = [1, 2, 3, missing]
+terrible_norm(@view(data[1:3]))
+```
+</div>
+
+**Output:**
+
+    MethodError: no method matching terrible_norm(::SubArray{Union{Missing, Int64},1,Array{Union{Missing, Int64},1},Tuple{UnitRange{Int64}},true})
+    Closest candidates are:
+      terrible_norm(!Matched::AbstractArray{#s6,1} where #s6<:Real) at In[36]:2
+
+Or from source that *could* contain non-Real values, but that actually doesn't
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+let
+    x = []
+    for ii in 1:10
+        push!(x, ii)
+    end
+    
+    terrible_norm(x)
+end
+
+```
+</div>
+
+**Output:**
+
+    MethodError: no method matching terrible_norm(::Array{Any, 1})
+    Closest candidates are:
+      terrible_norm(!Matched::AbstractArray{#s6,1} where #s6<:Real) at In[36]:2
+
+### Dispatching on  `Function`
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+apply_inner(func::Function, xss) = [[func(x) for x in xs] for xs in xss]
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+apply_inner (generic function with 1 method)
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+apply_inner(round, [[0.2, 0.9], [1.2, 1.3, 1.6]])
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+2-element Array{Array{Float64,1},1}:
+ [0.0, 1.0]
+ [1.0, 1.0, 2.0]
+```
+
+</div>
+
+But this doesn't work on callable objects that don't subtype `Function`.
+Which include constructors.
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+apply_inner(Float32, [[0.2, 0.9], [1.2, 1.3, 1.6]])
+```
+</div>
+
+**Output:**
+
+    MethodError: no method matching apply_inner(::Type{Float32}, ::Array{Array{Float64,1},1})
+    Closest candidates are:
+      apply_inner(!Matched::Function, ::Any) at In[74]:1
+
+
+A workaround to that is to use `Base.Callable` which is a `Union{Type, Function}` so does functions and constructors.
+However, this will still miss-out on other callable objects, like `DiffEqBase.ODESolution` and `Flux.Chain`.
+(As `Base.Callable` is not exported, it is also probably not part of the intended public API of Julia.)
```

Comment:
It sounds like that you're suggesting `Base.Callable` but it comes with all kinds of restrictions.  Is there a better suggestion?

Reply:
WIll clarify, it is not a work around, it is a lesser antipattern